Source/santad/EventProviders/SNTEndpointSecurityAuthorizer.mm

```diff
@@ -0,0 +1,142 @@
+    if (RESPONSE_VALID(returnAction)) {
+      es_auth_result_t authResult = ES_AUTH_RESULT_DENY;
+
+      switch (returnAction) {
+        case ACTION_RESPOND_ALLOW_COMPILER:
+          [self.compilerController setProcess:msg->event.exec.target->audit_token isCompiler:true];
+          OS_FALLTHROUGH;
+        case ACTION_RESPOND_ALLOW: authResult = ES_AUTH_RESULT_ALLOW; break;
+        default: break;
+      }
+
+      [self respondToMessage:msg
+              withAuthResult:authResult
+                   cacheable:(authResult == ES_AUTH_RESULT_ALLOW)];
+      return;
+    } else if (returnAction == ACTION_REQUEST_BINARY) {
+      // TODO(mlw): Look into caching a `Deferred<value>` to better prevent
+      // raciness of multiple threads checking the cache simultaneously.
+      // Also mitigates need to poll.
+      usleep(5000);
+    } else {
+      break;
+    }
+  }
+
+  self->_authResultCache->AddToCache(targetFile, ACTION_REQUEST_BINARY);
+
+  [self.execController validateExecEvent:msg
+                              postAction:^bool(santa_action_t action) {
+                                return [self postAction:action forMessage:msg];
+                              }];
+}
+
+- (void)handleMessage:(Message &&)esMsg {
+  if (unlikely(esMsg->event_type != ES_EVENT_TYPE_AUTH_EXEC)) {
+    // This is a programming error
+    LOGE(@"Atteempting to authorize a non-exec event");
+    [NSException raise:@"Invalid event type"
+                format:@"Authorizing unexpected event type: %d", esMsg->event_type];
+  }
+
+  if (![self.execController synchronousShouldProcessExecEvent:esMsg]) {
+    [self postAction:ACTION_RESPOND_DENY forMessage:esMsg];
+    return;
+  }
+
+  [self processMessage:std::move(esMsg)
+               handler:^(const Message &msg) {
+                 [self processMessage:msg];
+               }];
+}
+
+- (bool)postAction:(santa_action_t)action forMessage:(const Message &)esMsg {
+  es_auth_result_t authResult;
+
+  switch (action) {
+    case ACTION_RESPOND_ALLOW_COMPILER:
+      [self.compilerController setProcess:esMsg->event.exec.target->audit_token isCompiler:true];
+      OS_FALLTHROUGH;
+    case ACTION_RESPOND_ALLOW: authResult = ES_AUTH_RESULT_ALLOW; break;
+    case ACTION_RESPOND_DENY: authResult = ES_AUTH_RESULT_DENY; break;
+    default:
+      // This is a programming error. Bail.
+      LOGE(@"Invalid action for postAction, exiting.");
+      [NSException raise:@"Invalid post action" format:@"Invalid post action: %d", action];
+  }
+
+  self->_authResultCache->AddToCache(esMsg->event.exec.target->executable, action);
+
+  // Don't cache DENY results. Santa only flushes ES cache when a new DENY rule
+  // is received. If DENY results were cached and a rule update made the
+  // executable allowable, ES would continue to apply the DENY cached result.
+  return [self respondToMessage:esMsg
+                 withAuthResult:authResult
+                      cacheable:(authResult == ES_AUTH_RESULT_ALLOW)];
```

Comment:
The comment above this is a little unclear - is this _only_ avoiding ES caching? Currently Santa does cache DENY responses but only for a short period, which AuthResultCache seems to have re-implemented.

Reply:
Correct, only the ES cache. Updated the comment to hopefully be more clear.